platform/sarif.go

```diff
@@ -265,3 +287,110 @@ func ReportId(projectName string) string {
 func JobUrl() string {
 	return os.Getenv("QODANA_JOB_URL")
 }
+
+func getRuleDescription(report *sarif.Report, ruleId string) string {
+	for _, run := range report.Runs {
+		for _, extension := range run.Tool.Extensions {
+			for _, rule := range extension.Rules {
+				if rule.Id == ruleId {
+					return fmt.Sprintf("Short description: %v\nFull description: %v",
+						rule.ShortDescription.Text,
+						rule.FullDescription.Text)
+				}
+			}
+		}
+	}
+	return ""
+}
+
+// ProcessSarif concludes the result of analysis based on provided SARIF file
+// - can print problems to the output
+// - can create GitLab CodeQuality issues report
+// - can submit problems to BitBucket Code Insights
+func ProcessSarif(sarifPath, analysisId, reportUrl string, printProblems, codeClimate, codeInsights bool) {
+	newProblems := 0
+	s, err := ReadReport(sarifPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var codeClimateIssues []CCIssue
+	var codeInsightIssues []bbapi.ReportAnnotation
+	if printProblems {
+		EmptyMessage()
+	}
+	for _, run := range s.Runs {
+		for _, r := range run.Results {
+			ruleId := r.RuleId
+			message := r.Message.Text
+			baselineState := baselineStateEmpty
+			if r.BaselineState != nil {
+				baselineState = r.BaselineState.(string)
+			}
+			if baselineState == baselineStateNew || baselineState == baselineStateEmpty {
+				newProblems++
+			}
+			if len(r.Locations) > 0 && baselineState != baselineStateUnchanged {
+				if codeClimate {
+					codeClimateIssues = append(codeClimateIssues, sarifResultToCodeClimate(&r))
+				}
+				if codeInsights {
+					ruleDescription := getRuleDescription(s, ruleId)
+					codeInsightIssues = append(codeInsightIssues, buildAnnotation(&r, ruleDescription, reportUrl))
+				}
+				if printProblems {
+					printSarifProblem(&r, ruleId, message)
+				}
+			}
+		}
+	}
+	if codeClimate {
+		err = writeGlCodeQualityReport(codeClimateIssues, sarifPath)
+		if err != nil {
+			log.Warnf("Problems writing GitLab CodeQuality report: %v", err)
+		}
+	}
+	if codeInsights {
+		err = sendBitBucketReport(codeInsightIssues, s.Runs[0].Tool.Driver.FullName, reportUrl, "qodana-"+analysisId)
+		if err != nil {
+			log.Warnf("Problems sending BitBucket Code Insights report: %v", err)
+		}
+	}
+	if !IsContainer() {
+		if newProblems == 0 {
+			SuccessMessage(getProblemsFoundMessage(0))
+		} else {
+			ErrorMessage(getProblemsFoundMessage(newProblems))
+		}
+	}
+}
+
+// getFingerprint returns the fingerprint of the Qodana (or not) SARIF result.
+func getFingerprint(r *sarif.Result) string {
+	if r != nil && r.PartialFingerprints != nil {
+		fingerprint, ok := r.PartialFingerprints["equalIndicator/v2"]
+		if ok {
+			return fingerprint
+		} else {
+			fingerprint, ok = r.PartialFingerprints["equalIndicator/v1"]
+			if ok {
+				return fingerprint
+			}
+		}
+	}
+	return RunGUID()
```

Comment:
I think we should panic here - all our reports _must_ have the fingerprints, so something without them is either not our sarif, or we have a serious issue.